### PR TITLE
aws: Increase the default master instance size to reduce etcd timeouts

### PR DESCRIPTION
**Pull request by smarterclayton**
After experimenting with disks we determined that we were CPU starved, not IO starved (later comments).  This PR sets default master size to m4.xlarge which performs more consistently against 1.11 kube.

https://openshift-gce-devel.appspot.com/build/origin-ci-test/logs/release-openshift-origin-installer-e2e-aws-4.0/3155/ is representative, many timeouts.

The CPU use in 1.11 kube is excessive due to two bugs that will be fixed in the rebase, after which we can try stepping back down to m4.large.

### CODE REVIEW

**Comment by wking** [comment]
Previous bump was #737.  I'd floated 500 GiB in #844.

/lgtm
/hold

Do we need an exception for this, or is it a bugfix?  Also, do we need to stuff this into 0.10.0, or can it land in follow-up work?

**Comment by smarterclayton** [comment]
Exception granted, bug fix, release blocker.  I don't think it has to be in 0.10.0 unless everything magically is fixed, in which case we'll rebuild 0.10.0

**Comment by smarterclayton** [comment]
Ok, here's the latency distribution of error messages in the logs:

image

Red is another run, blue is this PR, x-axis is MS.

So we're still seeing substantially the same curve, but the outliers are a bit better.

**Comment by smarterclayton** [comment]
Retrying with 120gb/400iop/io1 drives

**Comment by wking** [comment]
All nice and green :).  Can you plot a pretty graph with the io1 results too, so we can ooh and ah?

**Comment by smarterclayton** [comment]
Guess what?!

io1 did nothing!!!!

image

**Comment by smarterclayton** [comment]
I love data

**Comment by smarterclayton** [comment]
For the curious here's the one liner I used on the etcd logs:

    cat ~/Downloads/kube-system_etcd-member-ip-10-0-* | sed -E -e 's/.*took too long \(([^\)]+)\).*/\1/;t;d' | sed -E -e 's/^([0-9]+)\.([0-9]{3})([0-9])*s/\1\2.\3ms/' | sed -e 's/ms//' | sort -n > /tmp/times3.txt

**Comment by smarterclayton** [comment]
Trying with m4.xlarge

**Comment by smarterclayton** [comment]
m4.xlarge (4 cpu, up from 2) definitely improved it:

image

I think we should bump the instance size for now until the rebase lands, at which point we can look at reducing it.  The two known CPU issues should be addressed in the rebase.

**Comment by smarterclayton** [comment]
We have 600 m4.xlarge quota in the ci cluster

**Comment by wking** [comment]
Want to go back to gp2?  And for type, hit the other locations like #765?

**Comment by smarterclayton** [comment]
Do we need a bigger bootstrap node?

**Comment by smarterclayton** [comment]
back to gp2

**Comment by wking** [comment]
shouldn't have etcd on the bootstrap node.  More CPU couldn't hurt, but I'd guess were generally network-limoted.

**Comment by wking** [comment]
Also [here][1] (currently shared by masters and workers) and [here][2].

[1]: https://github.com/openshift/installer/blob/5f357d21585d1895b8737df2759c53117a2b2b00/pkg/asset/machines/worker.go#L30
[2]: https://github.com/openshift/installer/blob/a63e50e966ad91934cd77137ee65245a62d6f133/docs/user/aws/limits.md#instance-limits

**Comment by wking** [comment]
Mitigates #1059.

**Comment by smarterclayton** [comment]
I’m halfway between having a default we test and just changing it in e2e setup.  Before, we said we wanted to be consistent.  Now we’re starting to bump a bit bigger which is inconvenient for users,  and I can’t really say which way it should fall.

**Comment by wking** [comment]
I'm fine either way, but I know changing defaults here will work, while I'm less certain about install-config changes before #792 (they *should* work before that, but I haven't checked).  I'll check shortly.

**Comment by wking** [comment]
See #1076 for a pre-requisite for the CI-only approach.

For the installer-default approach you have in this PR, I've pushed up 73fb8d039 (`git pull ssh://github.com/wking/openshift-installer.git aws-ec2-instance-type-patch` or similar and squash in) with the changes I asked for [here][1].  Pull in anything that looks useful to you.

[1]: https://github.com/openshift/installer/pull/1069#issuecomment-454645518

**Comment by DanyC97** [comment]
> The CPU use in 1.11 kube is excessive due to two bugs 

@smarterclayton being curious (by nature ;) ) would you mind sharing the 2 bugs you mentioned

**Comment by smarterclayton** [comment]
@wking why don't you open a PR with  https://github.com/openshift/installer/commit/73fb8d039476d96fe6654ed85b8c2d6e1e788da3 and we'll take yours.  

I slept on it, to get to the next gate I'm ok with 3x4 3x2 for the next few weeks and then we can reassess.

**Comment by smarterclayton** [comment]
@DanyC97 there's an OpenAPI bug where it refreshes every 30s unnecessarily which causes huge amounts of allocations and CPU waste, and then there's one for fixing admission credentials.  I don't have the numbers off.

**Comment by smarterclayton** [comment]
Nm, grabbed it and integrated it

File: pkg/asset/machines/master.go [hunk 1]
```diff
@@ -81,7 +82,7 @@ func (m *Master) Generate(dependencies asset.Parents) error {
 		if err != nil {
 			return errors.Wrap(err, "failed to create master machine objects")
 		}
-		aws.ConfigMasters(machines, ic.ObjectMeta.Name)
+		aws.ConfigMasters(machines, ic.ObjectMeta.Name, mpool.InstanceType)
```
Comment (abhinavdahiya) [inline pkg/asset/machines/master.go]: why is this required?

File: pkg/asset/machines/worker.go [hunk 2]
```diff
@@ -26,9 +26,7 @@ import (
 )
 
 func defaultAWSMachinePoolPlatform() awstypes.MachinePool {
-	return awstypes.MachinePool{
-		InstanceType: "m4.large",
-	}
+	return awstypes.MachinePool{}
```
Comment (abhinavdahiya) [inline pkg/asset/machines/worker.go]: why is this emptied ?

Comment (smarterclayton) [reply pkg/asset/machines/worker.go]: @wking ^

File: pkg/asset/machines/master.go (hunk 1 above)
Comment (smarterclayton) [reply pkg/asset/machines/master.go]: @wking ^ this is what I get :)

Comment (wking) [reply pkg/asset/machines/master.go]: > why is this required?

Ah, sorry about that.  Yeah, we can probably drop this, since `aws.provider` should be applying this for us.

File: pkg/asset/machines/worker.go (hunk 2 above)
Comment (wking) [reply pkg/asset/machines/worker.go]: > why is this emptied ?

Because we no longer have any default AWS configuration shared by both masters and workers.  They each have their own defaults for instance types (and, with #1079) for volumes, that get applied directly after `defaultAWSMachinePoolPlatform()` calls.  In fact, my preference would be to drop `defaultAWSMachinePoolPlatform` entirely in favor of an explicit `awstypes.MachinePool{}` for initializing AWS machine pools.

File: pkg/asset/machines/master.go (hunk 1 above)
Comment (wking) [reply pkg/asset/machines/master.go]: @smarterclayton, can you unwind is change with:

```diff
diff --git a/pkg/asset/machines/aws/machines.go b/pkg/asset/machines/aws/machines.go
index 865d40a..7e04a99 100644
--- a/pkg/asset/machines/aws/machines.go
+++ b/pkg/asset/machines/aws/machines.go
@@ -120,12 +120,10 @@ func tagsFromUserTags(clusterID, clusterName string, usertags map[string]string)
 	return tags, nil
 }
 
-// ConfigMasters sets the PublicIP flag, bumps the instance type, and
-// assigns a set of load balancers to the given machines
-func ConfigMasters(machines []clusterapi.Machine, clusterName string, instanceType string) {
+// ConfigMasters sets the PublicIP flag and assigns a set of load balancers to the given machines
+func ConfigMasters(machines []clusterapi.Machine, clusterName string) {
 	for _, machine := range machines {
 		providerSpec := machine.Spec.ProviderSpec.Value.Object.(*awsprovider.AWSMachineProviderConfig)
-		providerSpec.InstanceType = instanceType
 		providerSpec.PublicIP = pointer.BoolPtr(true)
 		providerSpec.LoadBalancers = []awsprovider.LoadBalancerReference{
 			{
diff --git a/pkg/asset/machines/master.go b/pkg/asset/machines/master.go
index f4c3806..2c782ac 100644
--- a/pkg/asset/machines/master.go
+++ b/pkg/asset/machines/master.go
@@ -82,7 +82,7 @@ func (m *Master) Generate(dependencies asset.Parents) error {
 		if err != nil {
 			return errors.Wrap(err, "failed to create master machine objects")
 		}
-		aws.ConfigMasters(machines, ic.ObjectMeta.Name, mpool.InstanceType)
+		aws.ConfigMasters(machines, ic.ObjectMeta.Name)
 
 		list := listFromMachines(machines)
 		raw, err := yaml.Marshal(list)
```

I'm testing that now to make sure it still works...

File: pkg/asset/machines/worker.go (hunk 2 above)
Comment (wking) [reply pkg/asset/machines/worker.go]: Ah, maybe this was working around #1076.

**Comment by smarterclayton** [comment]
I had a run with setting proper resource constraints on etcd and it didn't move the needle enough (4x improvement).  We need this as well.

image

Blue is baseline, red is moving etcd to have a 300m CPU request (https://github.com/openshift/machine-config-operator/pull/316), and yellow is doubling cores.  So it's a good tail and 

I updated with the comments from trevor, let me know what else needs to be done.

**Comment by wking** [comment]
nit: can we squash down to one commit?

File: pkg/asset/machines/aws/machines.go [hunk 3]
```diff
@@ -120,7 +120,8 @@ func tagsFromUserTags(clusterID, clusterName string, usertags map[string]string)
 	return tags, nil
 }
 
-// ConfigMasters sets the PublicIP flag and assigns a set of load balancers to the given machines
+// ConfigMasters sets the PublicIP flag, bumps the instance type, and
+// assigns a set of load balancers to the given machines
```
Comment (wking) [inline pkg/asset/machines/aws/machines.go]: No need to bump this comment, since we're no longer bumping `ConfigMasters`.

**Comment by wking** [comment]
/lgtm
/hold

@smarterclayton, I'm fine if you want to pull the hold yourself, or if you want to wait for @abhinavdahiya and/or @crawford.

**Comment by wking** [comment]
/hold cancel

I see @crawford is already here ;).

**Comment by abhinavdahiya** [comment]
IMO this change is not required for this PR https://github.com/openshift/installer/pull/1069/files#diff-df3202277ffea159153362ec095a8762

**Comment by smarterclayton** [comment]
I can nuke that

**Comment by wking** [comment]
> IMO this change is not required for this PR https://github.com/openshift/installer/pull/1069/files#diff-df3202277ffea159153362ec095a8762

I think we want that change, but I'm fine going either way to get this PR landed and hashing it our later.

**Comment by wking** [comment]
[e2e-aws][1]:

```
fail [k8s.io/kubernetes/test/e2e/framework/util.go:2355]: Expected error:
    <*errors.errorString | 0xc4215d4d30>: {
        s: "failed to get logs from pod-configmaps-679ec92b-19da-11e9-9cbd-0a58ac1041b1 for configmap-volume-test: an error on the server (\"unknown\") has prevented the request from succeeding (get pods pod-configmaps-679ec92b-19da-11e9-9cbd-0a58ac1041b1)",
    }
    failed to get logs from pod-configmaps-679ec92b-19da-11e9-9cbd-0a58ac1041b1 for configmap-volume-test: an error on the server ("unknown") has prevented the request from succeeding (get pods pod-configmaps-679ec92b-19da-11e9-9cbd-0a58ac1041b1)
not to have occurred

Jan 16 22:02:27.651 W persistentvolume=volume-idempotent-delete-6svns Error deleting EBS volume "vol-037c7a92290037b82" since volume is in "deleting" state count(1)

failed: (37.9s) 2019-01-16T22:02:43 "[sig-storage] ConfigMap should be consumable from pods in volume as non-root with defaultMode and fsGroup set [NodeFeature:FSGroup] [Suite:openshift/conformance/parallel] [Suite:k8s]"
```

and more.

/retest

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/1069/pull-ci-openshift-installer-master-e2e-aws/2964/build-log.txt

**Comment by wking** [comment]
[images][1]:

```
2019/01/17 01:07:20 Copied 0.20Mi of artifacts from release-latest to /logs/artifacts/release-latest
2019/01/17 01:07:26 Ran for 9m46s
error: could not run steps: test "release-latest" failed: pod release-latest was already deleted
```

/retest

[1]: https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_installer/1069/pull-ci-openshift-installer-master-images/2205/build-log.txt

**Comment by jeremyeder** [comment]
related: https://github.com/openshift/installer/pull/1087

**Comment by mykaul** [comment]
Any idea if we need something similar for libvirt based installation? I'm keeping (since it doesn't seem to be accepted) this small diff:
```
[ykaul@ykaul installer]$ git diff
diff --git a/pkg/asset/machines/libvirt/machines.go b/pkg/asset/machines/libvirt/machines.go
index 909e08a51..defa4ab36 100644
--- a/pkg/asset/machines/libvirt/machines.go
+++ b/pkg/asset/machines/libvirt/machines.go
@@ -64,8 +64,8 @@ func provider(clusterName string, networkInterfaceAddress string, platform *libv
                        APIVersion: "libvirtproviderconfig.k8s.io/v1alpha1",
                        Kind:       "LibvirtMachineProviderConfig",
                },
-               DomainMemory: 2048,
-               DomainVcpu:   2,
+               DomainMemory: 4096,
+               DomainVcpu:   4,
                Ignition: &libvirtprovider.Ignition{
                        UserDataSecret: userDataSecret,
                },
```

But it seems the above kinda proves my point? (Note - did not test for etcd timeout, but I am frequently failing deployment on timeouts...)

**Comment by wking** [comment]
> I'm keeping (since it doesn't seem to be accepted) this small diff...

No need for patching, since #785 set these up to allow environment-variable overrides.  I can't speak for the other maintainers, but personally I'd rather punt further tuning until after the origin rebase lands, since that's rumored to have some memory-usage improvements.
